packages/gateway/src/manager.ts

```diff
@@ -54,6 +55,160 @@ export function createGatewayManager(options: CreateGatewayManagerOptions): Gate
+        const sessionInfo = await gateway.resharding.getSessionInfo()
+        logger.warn(`[Resharding] Session info retrieved.`)
+        // Don't have enough identify limits to try resharding
+        if (sessionInfo.sessionStartLimit.remaining < sessionInfo.shards) return { needed: false, info: sessionInfo }
+        logger.warn(`[Resharding] Able to reshard, checking whether necessary now.`)
+
+        // 2500 is the max amount of guilds a single shard can handle
+        // 1000 is the amount of guilds discord uses to determine how many shards to recommend.
+        // This algo helps check if your bot has grown enough to reshard.
+        const percentage = ((2500 * sessionInfo.shards) / (gateway.totalShards * 1000)) * 100
+        // Less than necessary% being used so do nothing
+        if (percentage < gateway.resharding.shardsFullPercentage) return { needed: false, info: sessionInfo }
+        logger.warn(`[Resharding] Resharding is needed.`)
+
+        return { needed: true, info: sessionInfo }
+      },
+      async reshard(info) {
+        logger.warn(`[Resharding] Starting the reshard process. Previous total shards. ${gateway.totalShards}`)
+        // Set values on gateway
+        gateway.totalShards = info.shards
+        // Handles preparing mid sized bots for LBS
+        gateway.totalShards = gateway.calculateTotalShards()
+        // Set first shard id if provided in info
+        if (info.firstShardId) gateway.firstShardId = info.firstShardId
+        // Set last shard id if provided in info
+        if (info.lastShardId) gateway.lastShardId = info.lastShardId
+        logger.warn(`[Resharding] Starting the reshard process. New Total Shards. ${gateway.totalShards}`)
+
+        // Resetting buckets
+        gateway.buckets.clear()
+        // Refilling buckets with new values
+        gateway.prepareBuckets()
+
+        // SPREAD THIS OUT TO DIFFERENT WORKERS TO BEGIN STARTING UP
+        gateway.buckets.forEach(async (bucket, bucketId) => {
+          for (const worker of bucket.workers) {
+            for (const shardId of worker.queue) {
+              await gateway.resharding.tellWorkerToPrepare(worker.id, shardId, bucketId)
+            }
+          }
+        })
+      },
+      async tellWorkerToPrepare(workerId, shardId, bucketId) {
+        logger.warn(`[Resharding] Telling worker to prepare. Worker: ${workerId} | Shard: ${shardId} | Bucket: ${bucketId}`)
+        const shard = new Shard({
+          id: shardId,
+          connection: {
+            compress: gateway.compress,
+            intents: gateway.intents,
+            properties: gateway.properties,
+            token: gateway.token,
+            totalShards: gateway.totalShards,
+            url: gateway.url,
+            version: gateway.version,
+          },
+          // Ignore events until we are ready
+          events: {
+            async message(shard, payload) {
+              if (payload.t === 'READY') {
```

Comment:
```suggestion
          events: {
            async message(shard, payload) {
              // Ignore events until we are ready
              if (payload.t === 'READY') {
```

Reply:
I'm not entirely sure if this comment talks about gateway events or the events like messageCreate etc.

Reply:
Code is moved to somewhere else now so resolving